lib/test/cdo/test_active_job_backend.rb

```diff
@@ -0,0 +1,333 @@
+
+      error = assert_raises(RuntimeError) do
+        Cdo::ActiveJobBackend.restart_workers_internal(5, n_batches: 1)
+      end
+
+      assert_match(/delayed_job: ERROR/, error.message)
+    end
+  end
+
+  describe 'start_n_workers()' do
+    it 'invokes run_process with correct process names' do
+      Cdo::ActiveJobBackend::Command.any_instance.expects(:run_process).with('delayed_job.3', anything).once
+      Cdo::ActiveJobBackend::Command.any_instance.expects(:run_process).with('delayed_job.4', anything).once
+      Cdo::ActiveJobBackend.start_n_workers(2, initial_worker_index: 3)
+    end
+  end
+
+  describe 'Command' do
+    it 'invokes valid methods of superclass Delayed::Command' do
+      assert Delayed::Command.method_defined?(:run_process)
+    end
+  end
+
+  describe 'stop_workers()' do
+    before do
+      Cdo::ActiveJobBackend::ExistingWorkers.stubs(:ps).returns(ps_for_4_stale_workers)
+      @stale_worker_pids = [89890, 89892, 89894, 89936]
+    end
+
+    it 'SIGTERMs workers first' do
+      sequence = Mocha::Sequence.new('stop_workers')
+      Cdo::ActiveJobBackend.expects(:kill).times(4).with('TERM', anything).in_sequence(sequence)
+      Cdo::ActiveJobBackend::ExistingWorkers.stubs(:ps).returns(ps_for_no_workers).in_sequence(sequence)
+
+      Cdo::ActiveJobBackend.stop_workers(@stale_worker_pids, {})
+    end
+
+    it "SIGKILLs workers if they don't exit after SIGTERM" do
+      sequence = Mocha::Sequence.new('stop_workers')
+      Cdo::ActiveJobBackend.expects(:kill).times(4).with('TERM', anything).in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:kill).times(4).with('KILL', anything).in_sequence(sequence)
+      Cdo::ActiveJobBackend::ExistingWorkers.stubs(:ps).returns(ps_for_no_workers).in_sequence(sequence)
+
+      Cdo::ActiveJobBackend.stop_workers(@stale_worker_pids, {}, timeout_s: 1)
+    end
+
+    it "raises exception if workers don't exit after SIGKILL" do
+      sequence = Mocha::Sequence.new('stop_workers')
+      Cdo::ActiveJobBackend.expects(:kill).times(4).with('TERM', anything).in_sequence(sequence)
+      Cdo::ActiveJobBackend.expects(:kill).times(4).with('KILL', anything).in_sequence(sequence)
+
+      error = assert_raises(Timeout::Error) do
+        Cdo::ActiveJobBackend.stop_workers(@stale_worker_pids, {}, timeout_s: 1)
+      end
+
+      assert_match(/delayed_job: ERROR, not all delayed_job worker processes terminated/, error.message)
+    end
+
+    it 'cleans up pid files' do
+      Cdo::ActiveJobBackend.stubs(:kill)
+      Cdo::ActiveJobBackend.stubs(:wait_for_workers_to_exit)
+
+      pid_file = File.join(@pid_dir, 'delayed_job.1.pid')
+      File.write(pid_file, '1001')
+
+      Cdo::ActiveJobBackend.stop_workers([1001], {1001 => pid_file})
+      refute File.exist?(pid_file)
+    end
+  end
+
+  describe 'verify_num_workers_running!()' do
+    it 'succeeds when n_workers_to_start matches ps' do
+      assert_equal 5, Cdo::ActiveJobBackend.verify_num_workers_running!(4)
+    end
```

Comment:
Its firstly a bug in the test, it should be 5 and 5. verify_num_workers_running!() only cares if there's _fewer_ than planned workers running. More than expected to start is covered by verifying age of workers instead. Maybe it should be a warning?

Reply:
I think I got confused because the `before` step that sets up the 5 process mock is way at the top of the file and there was a closer (and in a different scope) 4 process mock.